### ISSUE TITLE
Service Discovery: Reduce number of packages that are dependent on our SD implementations (e.g K8s, Azure) to just main.

### ISSUE THREAD

**Issue by bwplotka** [prometheus/prometheus#7107]
## Proposal

Prometheus has official support for numerous discovery mechanisms by default.

Currently, those are in the form of separate packages: https://github.com/prometheus/prometheus/tree/0730d6eb74a2a6e268260a800f5de90958d0253a/discovery
Each of those implementations is pretty specific and depends on large (very often larger than Prometheus itself) dependant Go modules (e.g Azure, Google, AWS).

**That causes a few issues:**

* Big Prometheus binary size (not a big problem, but causes some issues for rate users who run Prom within IoT, e.g. in their fridge)
* It is blocking the reusability of many packages. Currently, many packages depend on `config` -> depends on discovery stuff. This means that many packages are affected by huge dependency baggage. This blocks external projects e.g Thanos or Cortex to reuse such packages does not want to have extra dependencies that cause de hell (e.g mismatching versions). 

All of the above also results in a blocker to extend official ones with custom SD plugins (e.g for docker).

Unfortunately on top of this there is overall go module problems described many times.

**The overall first step, we should think about now, is how we can reduce the number of packages that depend on SD implementation to just one (main).**

### Prometheus Discovery package dependency Inspection

`goda graph -short github.com/prometheus/prometheus/cmd/prometheus | dot -Tsvg -o graph.svg`

NOTE: It's too big to render readable PNG Only SVG is usable. But just look on this! 

![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/6950331/78891355-554fdb00-7a5f-11ea-912c-48ba7cedabc5.png)

Most of it is Kubernetes, so after we ignore k8s to see anything and focusing on discovery:

`goda graph -short github.com/prometheus/prometheus/cmd/prometheus - github.com/prometheus/prometheus/discovery/kubernetes | dot -Tsvg -o smgraph.svg`

![ezgif com-crop (4)](https://user-images.githubusercontent.com/6950331/78892753-d90ac700-7a61-11ea-829a-14be5c25fbf0.png)

As we see there are 3 links we might want to change as part of this work:

* [ ] `main` -> `discovery` -> `discovery/config` -> SD implementations
* [ ] `main` -> `discovery/config` -> SD implementations
* [ ] `main` -> `config` ->  `discovery/config` -> SD implementations

Idea is to have just `main`. The problem is how to make the indirection required for `config`. Well, there is one solution that we chosen in Thanos, see:

https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/objstore/client/factory.go#L38

With such factory and config alignment we can have instead of this:

```
// ScrapeConfig configures a scraping unit for Prometheus.
type ScrapeConfig struct {
         // ( ...)

	ServiceDiscoveryConfig sd_config.ServiceDiscoveryConfig `yaml:",inline"`
}
```
We can have:

```
// ScrapeConfig configures a scraping unit for Prometheus.
type ScrapeConfig struct {
         // ( ...) b
	ServiceDiscoveryConfig yaml.Unmarshaler `yaml:",inline"`
}
```

Then we can have unmarshaler which knows which implementation's configuration to use. Such marshaller will be added by main.go only, so config will no longer be dependent on all SD. 

Help wanted! :hugs:

**Comment by roidelapluie** [comment]
I removed low hanging fruit as I think that it should be done by someone experienced; I think marking it as low hanging fruit might lead to pull requests that will take ages to be merged. WDYT?

**Comment by brian-brazil** [comment]
Yeah, this one is going to be non-trivial. It's not the sort of things someone new to the project could plausibly figure out in an afternoon.

**Comment by abursavich** [comment]
I'm interested in taking a crack at this, but by definition it's going to require breaking package APIs... Are there any stability guarantees around the config/discovery packages that must be respected, besides the YAML formatting?

**Comment by roidelapluie** [comment]
> I'm interested in taking a crack at this, but by definition it's going to require breaking package APIs... Are there any stability guarantees around the config/discovery packages that must be respected, besides the YAML formatting?

Hello, I am already working on this one.

**Comment by abursavich** [comment]
> Hello, I am already working on this one.

I see. Here's what I had already implemented: https://gist.github.com/abursavich/dfd8541dad8e33e1eeb12af987e1a58d

**Comment by abursavich** [comment]
Sure.

**Comment by brian-brazil** [comment]
> Are there any stability guarantees around the config/discovery packages that must be respected, besides the YAML formatting?

There are none, however we should try to avoid requiring any changes to https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd
If we do end up having to break that, there's some more things I'd like to cleanup there while we're at it.

**Comment by abursavich** [comment]
I don't see anything in `custom-sd` that would need to change.

**Comment by abursavich** [comment]
In an attempt to make this as painless as possible, I've gone down a pretty deep rabbit hole with YAML parsing... creating several issues and PRs upsteam (https://github.com/go-yaml/yaml/issues/639, https://github.com/go-yaml/yaml/issues/640, https://github.com/go-yaml/yaml/issues/641, https://github.com/go-yaml/yaml/issues/642, https://github.com/go-yaml/yaml/issues/643).

Implementing this change requires using a custom yaml.Unmarshaler for the type that will hold the service discovery configs (currently ServiceDiscoveryConfig), since that type can't have dependencies on the implementations and the expected fields will be dynamic at runtime. These configs are then inlined into ScrapeConfig and AlertmanagerConfig, which is where we run into a problem with using "strict" yaml unmarshaling.

Both yaml.v2 and yaml.v3 rely solely on the fields in the go structs to implement known-field matching. Currently, there's no way to account for fields that are only used in a custom yaml.Unmarshaler and do not otherwise exist in the structs. These "extra" fields that the custom yaml.Unmarshaller is actually expecting will always result in an "unexpected field" error.

Here's a small-ish example: https://play.golang.com/p/Ma_TavWqKwI

Without making deep (unwelcome?) changes to the yaml package, it seems like the options are:
1. Remove strictness in unmarshaling yaml configs. Configs with typos or unknown/unsupported fields would be silently accepted.
2. Remove inlining of service discovery configs in ScrapeConfig and AlertmanagerConfig. This would be a breaking change and existing YAML configs would need to be modified when upgrading.
3. Hoist the dynamic unmarshaling out of a shared type (e.g. ServiceDiscoveryConfig) and implement it in both ScrapeConfig and AlertmanagerConfig's custom UnmarshalYAML methods. This will make the implementation more complicated and messy, but should have the least impact.

Thoughts before proceeding further?

**Comment by roidelapluie** [comment]
More path:

1. The approach used in https://github.com/Ompluscator/dynamic-struct

Or even unmarshalling a map[string]interface{}, removing the keys from SD's and re-marshalling the rest of the struct.

**Comment by abursavich** [comment]
I hadn't seen dynamic-struct package before, but I was using reflect the same way to create a new type with fields for the registered discovery mechanisms at runtime.  The problem is that I can't (3) abstract the special yaml.Unmarshaler into a single static type, (2) inline that type into other types (ScrapeConfig and AlertmanagerConfig), and (1) support strict decoding mode. I have to give up on one of 1, 2, or 3.

**Comment by roidelapluie** [comment]
1. and 2. are not acceptable to me:

1. leads users with issues like typos etc which means that Prometheus does not behave like they would expect (e.g. because they use scrape config).
2. This is breaking change too; unless we hardcode all the existing implementations (?) but that does not sound right.

I'd say that we should go for 3. I am not afraid of it being "a bit more messy" if it's properly tested. Do you already have an idea of what it would look like?

(or ro you think we could also upstram to yaml an option to register fields to ignore with the strict unmarshaling?)

**Comment by abursavich** [comment]
I'm also leaning to 3. I haven't written anything for it yet, but I'll try to come up with something clean. I don't think this will be a problem in practice, but it's worth mentioning that the same problem will reemerge if ScrapeConfig or AlertmanagerConfig are ever inlined into any other config types.

**Comment by roidelapluie** [comment]
Awesome, should you be blocked or have further questions, do not hesitate to come back to us.

Note that if you think that 2. would be really a lot more clean, maybe we could keep a compatibility layer for the existing service discoveries and add the new ones to the new way of configuring SD's.

**Comment by abursavich** [comment]
Cool.

I think 2 would be a very bad user experience, but it would be the easiest way of side-stepping the issues with the yaml package now and in the future. If it would require adding a compatibility layer, then I don't think there's any reason to do it. I mostly bring it up for completeness and in case there's another breaking change in the works and it makes sense to bundle them together.

**Comment by roidelapluie** [comment]
Actually, we could just keep the field names as it for now. Without having two ways to configure them.

And just users would not be able to register *any* service discoveries, just the ones we have now.

In the mean time we can see with upstream yaml what would be the best solution.

**Comment by roidelapluie** [comment]
Also, it 3 is just having a field: CustomSDs map[string]interface{} "yaml,inline" and raise an error upon Unmarshal when the sd is not known, that is probably not too hard to implement?

**Comment by abursavich** [comment]
Fixed by #7701.
